realm/src/main/java/io/realm/RealmObject.java

```diff
@@ -124,4 +135,142 @@ protected static Realm getRealm(RealmObject obj) {
     protected static Row getRow(RealmObject obj) {
         return obj.row;
     }
+
+    /**
+     * Set the Future instance returned by the worker thread, we need this instance
+     * to force {@link #load()} an async query, we use it to determine if the current
+     * RealmResults is a sync or async one.
+     *
+     * @param pendingQuery pending query.
+     */
+    void setPendingQuery(Future<Long> pendingQuery) {
+        this.pendingQuery = pendingQuery;
+        if (isLoaded()) {
+            // the query completed before RealmQuery
+            // had a chance to call setPendingQuery to register the pendingQuery (used btw
+            // to determine isLoaded behaviour)
+            onCompleted();
+
+        } // else, it will be handled by the Realm#handler
+    }
+
+    /**
+     * Determine if the current RealmObject is obtained synchronously or asynchronously (from
+     * a worker thread). Synchronous RealmObjects are by definition blocking hence this method
+     * will always return {@code true} for them.
+     * Note: This will return {@code true} if called from a standalone object (created outside of Realm).
+     *
+     * @return {@code true} if the query has completed and the data is available {@code false} if the
+     * query is in progress.
+     */
+    public boolean isLoaded() {
+        if (realm == null) {
+            return true;
+        }
+        realm.checkIfValid();
+        return pendingQuery == null || isCompleted;
+    }
+
+    /**
+     * Make an asynchronous query blocking.
+     * Note: This will return {@code true} if called from a standalone object (created outside of Realm).
```

Comment:
Any code that just accepts a RealmObject without caring where it comes from. Throwing an exception would be overly restrictive IMO:

```
public void method(Foo foo) {
  if (foo.isLoaded()) {
    showFoo(foo);
  } else {
    foo.addChangeListener(() -> showFoo(foo));
  }
}
```